### ISSUE TITLE
LES client consumes 100% of a CPU core when DNS discovery enabled

### ISSUE THREAD

**Issue by veox** [ethereum/go-ethereum#22509]
### System information

Geth version: `geth version`

```
Geth
Version: 1.10.1-stable
Git Commit: c2d2f4ed8f232bb11663a1b01a2e578aa22f24bd
Architecture: amd64
Go Version: go1.16
Operating System: linux
GOPATH=
GOROOT=go
```

OS & Version: Linux, Ubuntu 16.04.7 LTS (Xenial Xerus)

Commit hash: N/A (installed through official PPA)

### Expected behaviour

Same as 1.9.x branch behaved - about 5% to 10% of a CPU consumed.

### Actual behaviour

100% of a core is consumed.

Note that this is intentionally a low-resource VM, and it only has 1 core total.

### Steps to reproduce the behaviour

Run as:

```
nice ionice -c3 /usr/bin/geth --syncmode "light" --datadir /home/geth/.ethereum --cache 1024 --maxpeers 100
```

I'm actually doing this through a `systemd` unit, but the above is a succinct one-liner.

### CPU profiling graph

(Sorry for the PNG, github got picky about the SVG. :/)

![karl-100cpu prof](https://user-images.githubusercontent.com/3036030/111369990-58f0d680-86a0-11eb-8930-d8e50276a998.png)

The "culprit" is `Now()->nanotime()`, and the code path leading through DNS discovery.

If I had to guess, the issue stems from there not being enough peers (LES servers) in the network - neither the default of 25, nor the desirable 100, so the path thrashes while looking for more. The buck stops at the clock because "it's not time to check yet".

Possibly related based on title: issue #22306

### Workaround

Can confirm that disabling DNS discovery with

`--discovery.dns ""`

removes the issue.

CPU graph pre-upgrade (up to 19:50), after upgrade (19:50-21:30), after workaround (after 21:30):

![image](https://user-images.githubusercontent.com/3036030/111370902-72465280-86a1-11eb-927c-bb6d667ff21d.png)

**Comment by holiman** [comment]
Thanks for the very detailed report!

**Comment by fjl** [comment]
After some debugging, we now know why this happens:

In Geth v1.10, we changed the structure of the "les" ENR entry. As a result, the DHT crawler that creates the DNS lists 
no longer recognizes the les nodes (this is fixed in #22565). So at the moment, the les DNS list only contains non-upgraded nodes which are invalid according to the ForkID filter.

When the light client searches for servers, it combines the results of the DHT search and DNS discovery iterator using FairMix. It filters the output of this combined stream of nodes according to the ForkID. 

The DNS node iterator caches previously resolved records in an LRU cache. Since the les node lists don't contain a lot of nodes, the whole list can be cached and getting the 'next node' does not incur any network-based delay. So the les discovery just keeps iterating the list of incompatible nodes over and over in a hot loop.

I'm still unsure how to best fix this. It seems that the best fix would be to speed-limit the DNS iterator even when the response comes from the cache.
